piptools/writer.py

```diff
@@ -257,14 +260,25 @@ def write(
         hashes: Optional[Dict[InstallRequirement, Set[str]]],
     ) -> None:
 
-        for line in self._iter_lines(results, unsafe_requirements, markers, hashes):
-            if self.dry_run:
-                # Bypass the log level to always print this during a dry run
-                log.log(line)
-            else:
-                log.info(line)
-                self.dst_file.write(unstyle(line).encode())
-                self.dst_file.write(os.linesep.encode())
+        if not self.dry_run:
```

Comment:
FWIW this could be cleaner if implemented with a context manager conditionally returning a dummy I/O object.

Reply:
this refactor requires `from __future__ import annotations` to get flake8 to pass: https://github.com/jazzband/pip-tools/pull/1696